clippy_lints/src/unused_self.rs

```diff
@@ -0,0 +1,102 @@
+use if_chain::if_chain;
+use rustc::hir::def::Res;
+use rustc::hir::intravisit::{walk_path, NestedVisitorMap, Visitor};
+use rustc::hir::{AssocItemKind, HirId, ImplItemKind, ImplItemRef, Item, ItemKind, Path};
+use rustc::lint::{LateContext, LateLintPass, LintArray, LintPass};
+use rustc::{declare_lint_pass, declare_tool_lint};
+
+use crate::utils::span_help_and_lint;
+
+declare_clippy_lint! {
+    /// **What it does:** Checks methods that contain a `self` argument but don't use it
+    ///
+    /// **Why is this bad?** It may be clearer to define the method as a static function instead
+    /// of an instance method if it doesn't require `self`.
+    ///
+    /// **Known problems:** None.
+    ///
+    /// **Example:**
+    /// ```rust,ignore
+    /// struct A;
+    /// impl A {
+    ///     fn method(&self) {}
+    /// }
+    /// ```
+    ///
+    /// Could be written:
+    ///
+    /// ```rust,ignore
+    /// struct A;
+    /// impl A {
+    ///     fn method() {}
+    /// }
+    /// ```
+    pub UNUSED_SELF,
+    complexity,
```

Comment:
I'd rather make this a pedantic lint. This is because of three reasons:

1. I can see that people prefer to have a `self.func()` call instead of a `Self::func()` call. Less characters to write and more familiar in rust code.
2. Similar to 1.: If you access this function from outside the impl, you may have to write `LongDescriptiveStructName::func()` instead of `ldsn.func()` (or another short name for variable). 
3. On a function `fn func(self)`, where `self` isn't `Copy` it may be intended to move `self` in the function and drop it afterwards. [Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=a02ce3a0fa90f061e001d9b51941d6ed)